### PR TITLE
Add a script to update MLflow package versions

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Add a script to update MLflow package versions. Our internal CI job uses this script.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

### Before release

```
python dev/update_mlflow_versions.py before-release --new-version 1.29.0
```

```diff
diff --git a/mlflow/R/mlflow/DESCRIPTION b/mlflow/R/mlflow/DESCRIPTION
index 4dfab08c3..9bbb04c59 100644
--- a/mlflow/R/mlflow/DESCRIPTION
+++ b/mlflow/R/mlflow/DESCRIPTION
@@ -1,7 +1,7 @@
 Type: Package
 Package: mlflow
 Title: Interface to 'MLflow'
-Version: 1.28.1
+Version: 1.29.0
 Authors@R: 
     c(person(given = "Matei",
              family = "Zaharia",
diff --git a/mlflow/java/client/pom.xml b/mlflow/java/client/pom.xml
index 7d18c2272..56dac7eb5 100644
--- a/mlflow/java/client/pom.xml
+++ b/mlflow/java/client/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.mlflow</groupId>
     <artifactId>mlflow-parent</artifactId>
-    <version>1.28.1-SNAPSHOT</version>
+    <version>1.29.0</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/mlflow/java/pom.xml b/mlflow/java/pom.xml
index b9c260929..0b7c57bec 100644
--- a/mlflow/java/pom.xml
+++ b/mlflow/java/pom.xml
@@ -3,7 +3,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>org.mlflow</groupId>
   <artifactId>mlflow-parent</artifactId>
-  <version>1.28.1-SNAPSHOT</version>
+  <version>1.29.0</version>
   <packaging>pom</packaging>
   <name>MLflow Parent POM</name>
   <url>http://mlflow.org</url>
@@ -59,7 +59,7 @@
   </distributionManagement>
 
   <properties>
-    <mlflow-version>1.28.1-SNAPSHOT</mlflow-version>
+    <mlflow-version>1.29.0</mlflow-version>
     <maven.compiler.source>1.8</maven.compiler.source>
     <maven.compiler.target>1.8</maven.compiler.target>
     <scala.version>2.11.12</scala.version>
diff --git a/mlflow/java/scoring/pom.xml b/mlflow/java/scoring/pom.xml
index 14fea1100..87495cff7 100644
--- a/mlflow/java/scoring/pom.xml
+++ b/mlflow/java/scoring/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.mlflow</groupId>
     <artifactId>mlflow-parent</artifactId>
-    <version>1.28.1-SNAPSHOT</version>
+    <version>1.29.0</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/mlflow/java/spark/pom.xml b/mlflow/java/spark/pom.xml
index 15ad5098c..b2ded5134 100644
--- a/mlflow/java/spark/pom.xml
+++ b/mlflow/java/spark/pom.xml
@@ -1,7 +1,7 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
   <modelVersion>4.0.0</modelVersion>
   <artifactId>mlflow-spark</artifactId>
-  <version>1.28.1-SNAPSHOT</version>
+  <version>1.29.0</version>
   <name>${project.artifactId}</name>
   <properties>
     <maven.compiler.source>1.8</maven.compiler.source>
@@ -15,7 +15,7 @@
   <parent>
     <groupId>org.mlflow</groupId>
     <artifactId>mlflow-parent</artifactId>
-    <version>1.28.1-SNAPSHOT</version>
+    <version>1.29.0</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/mlflow/server/js/src/common/constants.js b/mlflow/server/js/src/common/constants.js
index f2f100671..df3edb1f9 100644
--- a/mlflow/server/js/src/common/constants.js
+++ b/mlflow/server/js/src/common/constants.js
@@ -8,7 +8,7 @@ export const ErrorCodes = {
   PERMISSION_DENIED: 'PERMISSION_DENIED',
 };
 
-export const Version = '1.28.1.dev0';
+export const Version = '1.29.0';
 
 const DOCS_VERSION = 'latest';
 
diff --git a/mlflow/version.py b/mlflow/version.py
index 1a21248fa..0ee5f3d9f 100644
--- a/mlflow/version.py
+++ b/mlflow/version.py
@@ -2,7 +2,7 @@
 import re
 
 
-VERSION = "1.28.1.dev0"
+VERSION = "1.29.0"
 
 
 def is_release_version():
```

### After release

```
python dev/update_mlflow_versions.py after-release --new-version 1.29.0
```

```diff
diff --git a/mlflow/R/mlflow/DESCRIPTION b/mlflow/R/mlflow/DESCRIPTION
index 4dfab08c3..e2607f00e 100644
--- a/mlflow/R/mlflow/DESCRIPTION
+++ b/mlflow/R/mlflow/DESCRIPTION
@@ -1,7 +1,7 @@
 Type: Package
 Package: mlflow
 Title: Interface to 'MLflow'
-Version: 1.28.1
+Version: 1.29.1
 Authors@R: 
     c(person(given = "Matei",
              family = "Zaharia",
diff --git a/mlflow/java/client/pom.xml b/mlflow/java/client/pom.xml
index 7d18c2272..a6bd4700e 100644
--- a/mlflow/java/client/pom.xml
+++ b/mlflow/java/client/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.mlflow</groupId>
     <artifactId>mlflow-parent</artifactId>
-    <version>1.28.1-SNAPSHOT</version>
+    <version>1.29.1-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/mlflow/java/pom.xml b/mlflow/java/pom.xml
index b9c260929..c3bf65742 100644
--- a/mlflow/java/pom.xml
+++ b/mlflow/java/pom.xml
@@ -3,7 +3,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>org.mlflow</groupId>
   <artifactId>mlflow-parent</artifactId>
-  <version>1.28.1-SNAPSHOT</version>
+  <version>1.29.1-SNAPSHOT</version>
   <packaging>pom</packaging>
   <name>MLflow Parent POM</name>
   <url>http://mlflow.org</url>
@@ -59,7 +59,7 @@
   </distributionManagement>
 
   <properties>
-    <mlflow-version>1.28.1-SNAPSHOT</mlflow-version>
+    <mlflow-version>1.29.1-SNAPSHOT</mlflow-version>
     <maven.compiler.source>1.8</maven.compiler.source>
     <maven.compiler.target>1.8</maven.compiler.target>
     <scala.version>2.11.12</scala.version>
diff --git a/mlflow/java/scoring/pom.xml b/mlflow/java/scoring/pom.xml
index 14fea1100..436b36757 100644
--- a/mlflow/java/scoring/pom.xml
+++ b/mlflow/java/scoring/pom.xml
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.mlflow</groupId>
     <artifactId>mlflow-parent</artifactId>
-    <version>1.28.1-SNAPSHOT</version>
+    <version>1.29.1-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/mlflow/java/spark/pom.xml b/mlflow/java/spark/pom.xml
index 15ad5098c..4a54e492f 100644
--- a/mlflow/java/spark/pom.xml
+++ b/mlflow/java/spark/pom.xml
@@ -1,7 +1,7 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
   <modelVersion>4.0.0</modelVersion>
   <artifactId>mlflow-spark</artifactId>
-  <version>1.28.1-SNAPSHOT</version>
+  <version>1.29.1-SNAPSHOT</version>
   <name>${project.artifactId}</name>
   <properties>
     <maven.compiler.source>1.8</maven.compiler.source>
@@ -15,7 +15,7 @@
   <parent>
     <groupId>org.mlflow</groupId>
     <artifactId>mlflow-parent</artifactId>
-    <version>1.28.1-SNAPSHOT</version>
+    <version>1.29.1-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/mlflow/server/js/src/common/constants.js b/mlflow/server/js/src/common/constants.js
index f2f100671..e28d9f2d1 100644
--- a/mlflow/server/js/src/common/constants.js
+++ b/mlflow/server/js/src/common/constants.js
@@ -8,7 +8,7 @@ export const ErrorCodes = {
   PERMISSION_DENIED: 'PERMISSION_DENIED',
 };
 
-export const Version = '1.28.1.dev0';
+export const Version = '1.29.1.dev0';
 
 const DOCS_VERSION = 'latest';
 
diff --git a/mlflow/version.py b/mlflow/version.py
index 1a21248fa..e7e36c4fe 100644
--- a/mlflow/version.py
+++ b/mlflow/version.py
@@ -2,7 +2,7 @@
 import re
 
 
-VERSION = "1.28.1.dev0"
+VERSION = "1.29.1.dev0"
 
 
 def is_release_version():

```

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: dev/update_mlflow_versions.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from typing import List
+
+import click
+from packaging.version import Version
+
+
+def get_current_version() -> str:
+    text = Path("mlflow", "version.py").read_text()
+    ver = Version(re.search(r'VERSION = "(.+)"', text).group(1))
+    return f"{ver.major}.{ver.minor}.{ver.micro}"
```
Comment (BenWilson2) [inline dev/update_mlflow_versions.py:12]: Could the `.public` property be used here?

```python
return ver.public
```

File: dev/update_mlflow_versions.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+Update MLflow package versions BEFORE release.
+
+Usage:
+
+python dev/update_mlflow_versions.py before-release --new-version 1.29.0
+"""
+)
+@click.option("--new-version", required=True, help="New version to release")
+def before_release(new_version: str):
+    update_versions(new_version, add_dev_suffix=False)
```
Comment (BenWilson2) [inline dev/update_mlflow_versions.py:80]: Would it be helpful to have a simple check that if the `new_version` specified is less than or equal to the current version in the rc branch that a click exception is thrown?

**Review by BenWilson2 — APPROVED** [review]
LGTM; just 2 nit picky questions

File: dev/update_mlflow_versions.py (hunk 1 above)
Comment (harupy) [reply dev/update_mlflow_versions.py:12]:
```
>>> Version("1.28.1.dev0").public
'1.28.1.dev0'
```

The `public` property ([implementation](https://github.com/pypa/packaging/blob/8c6a45e20857db912d397afb489832d9528a84c8/packaging/version.py#L354)) doesn't seem to drop the dev suffix.

File: dev/update_mlflow_versions.py (hunk 2 above)
Comment (harupy) [reply dev/update_mlflow_versions.py:80]: Sounds good!

File: dev/update_mlflow_versions.py (hunk 1 above)
Comment (BenWilson2) [reply dev/update_mlflow_versions.py:12]: That's right... I forgot that we don't use post markings for dev versions (`+`).

**Review by BenWilson2 — APPROVED** [review]
Great check added! LGTM! :)

**Pull request merged**